src/components/TreeSelect.tsx

```diff
@@ -0,0 +1,337 @@
+          resultState: state.filter(v => v !== value && v !== parentValue),
+          tData
+        })
+      );
+    } else {
+      let siblingsChecked = true;
+      siblings.forEach(sibling => {
+        siblingsChecked = siblingsChecked && state.includes(sibling.value);
+      });
+      const shouldCheckParent = parentValue && siblingsChecked;
+      // use set in case parent.value already exists in state
+      const resultState = Array.from(
+        new Set(
+          [...state, value].concat(shouldCheckParent ? [parentValue] : [])
+        )
+      );
+      onChange(adjustAll({ resultState, tData }));
+    }
+  }
+};
+
+// selects or deselects the All checkbox depending on current options
+const adjustAll = ({
+  resultState,
+  tData
+}: {
+  resultState: string[];
+  tData: TreeDataEntry[];
+}) => {
+  const allValues = getAllValues(tData).filter(value => value !== ALL_VALUE);
+  const resultStateHasAllValues = allValues.reduce(
+    (accum, value) => accum && resultState.includes(value),
+    true
+  );
+  // convert to set in case all exists in URL when its not supposed to
+  const resultStateSet = new Set(resultState);
+  if (resultStateHasAllValues) {
+    resultStateSet.add(ALL_VALUE);
+  } else {
+    resultStateSet.delete(ALL_VALUE);
+  }
+  return Array.from(resultStateSet);
+};
+
+interface FindNodeResult {
+  target: TreeDataEntry;
+  parent: TreeDataEntry;
+  siblings: TreeDataEntry[] | TreeDataChildEntry[];
+}
+
+const findNode = ({
+  value,
+  tData
+}: {
+  value: string;
+  tData: TreeDataEntry[];
+}): FindNodeResult => {
+  for (const curr of tData) {
+    if (curr.value === value) {
+      return {
+        target: curr,
+        parent: null,
+        siblings: tData.filter(v => v.value !== value)
+      };
+    }
+    if (curr.children) {
+      const child = curr.children.find(c => c.value === value);
+      if (child) {
+        return {
+          target: child,
+          parent: curr,
+          siblings: curr.children.filter(c => c.value !== value)
+        };
+      }
+    }
+  }
+  return {
+    target: null,
+    parent: null,
+    siblings: []
+  };
+};
+
+// returns all values in tData from parents and children
+const getAllValues = (tData: TreeDataEntry[]): string[] => {
+  return tData.reduce((accum, currNode) => {
+    const childrenValues = currNode.children
+      ? currNode.children.map(child => child.value)
+      : [];
+    return accum.concat([currNode.value]).concat(childrenValues);
+  }, []);
+};
+
+// depth first traversal checkbox data.
+// pushes parent then children to rows array
+// keeps track of level for indentation
+const renderCheckboxes = ({
+  tData,
+  state,
+  onChange
+}: {
+  tData: TreeDataEntry[];
+  state: string[];
+  onChange: (v: [string]) => void;
+}) => {
+  const rows: JSX.Element[] = [];
+  tData.forEach(entry => {
+    renderCheckboxesHelper({ rows, data: entry, onChange, state, tData });
+  });
+  return rows;
+};
+
+const renderCheckboxesHelper = ({
```

Comment:
we talked offline about render functions and both agree that they have purpose and value but that components will lead to better abstractions. this code is fine as is and we can implement our decision around render funcs going forward.